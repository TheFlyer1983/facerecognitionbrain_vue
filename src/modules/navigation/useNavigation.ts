import {
  RouteLocationRaw,
  useRouter
} from 'vue-router';

function useNavigation() {
  const router = useRouter();

  function navigate(route: RouteLocationRaw) {
    router.push(route);
  }

  return { navigate };
}

export default useNavigation;
