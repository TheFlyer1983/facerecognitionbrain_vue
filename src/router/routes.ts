export enum Routes {
  Home = 'home'
}