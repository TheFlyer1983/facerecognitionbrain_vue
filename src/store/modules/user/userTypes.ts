import { ActionContext } from 'vuex';

export type UserActionContext = ActionContext<UserState, {}>

export interface UserState {
  isSignedIn: boolean;
  isProfileOpen: false;
  user: User | null;
  token: string;
  rank: string;
}

export interface User {
  id: string;
  name: string;
  email: string;
  entries: number;
  joined: string;
  pet: string;
  age: string;
}
