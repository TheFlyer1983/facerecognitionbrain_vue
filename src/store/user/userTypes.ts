export interface UserState {
  isSignedIn: boolean;
  isProfileOpen: boolean;
  id: string;
  user: User | null;
  token: string;
  rank: string;
}

export interface User {
  name: string;
  email: string;
  entries: number;
  joined: string | number | Date;
  pet?: string;
  age?: number;
  role?: string;
}

export interface UpdateInfo {
  name?: string;
  pet?: string;
  age?: number;
}
