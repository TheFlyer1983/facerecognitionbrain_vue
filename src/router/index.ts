import { createRouter, createWebHistory, RouteRecordRaw } from 'vue-router';

import HelloWorld from '@/views/HelloWorld.vue';
import { Routes } from './routes';

export const INITIAL_ROUTE = '/';

const routes: Array<RouteRecordRaw> = [
  {
    path: INITIAL_ROUTE,
    name: 'Index',
    redirect: () => ({ name: Routes.Home })
  },
  {
    path: '/home',
    name: Routes.Home,
    component: HelloWorld,
    props: { msg: 'Hello Vue 3 + TypeScript + Vite' }
  },
  {
    path: '/login',
    name: Routes.Login,
    component: () => import('@/views/Login/Login.vue')
  },
  {
    path: '/register',
    name: Routes.Register,
    component: () => import('@/views/Register/Register.vue')
  }
];

const router = createRouter({
  history: createWebHistory(INITIAL_ROUTE),
  routes
});

export default router;
