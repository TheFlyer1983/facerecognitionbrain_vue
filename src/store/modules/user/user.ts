import { UserState, UserActionContext } from './userTypes';

import { endpoints } from '@/constants';
import { LoginInfo, RegisterInfo } from '@/types';
import {
  getAuthTokenInSession,
  saveAuthTokenInSession
} from '@/functions/storageFunctions';

export const state: UserState = {
  isSignedIn: false,
  isProfileOpen: false,
  user: null,
  token: ''
};

export const getters = {
  getUser(state: UserState): UserState['user'] {
    return state.user;
  },
  getToken(state: UserState): UserState['token'] {
    return state.token;
  },
  getIsSignedIn(state: UserState): UserState['isSignedIn'] {
    return state.isSignedIn;
  }
};

export const mutations = {
  setUserId(state: UserState, payload: string): void {
    state.user = { ...state.user, id: payload } as UserState['user'];
  },

  setUser(state: UserState, payload: UserState['user']): void {
    state.user = payload;
  },

  toggleSignIn(state: UserState, payload: UserState['isSignedIn']): void {
    state.isSignedIn = payload;
  },

  setToken(state: UserState, payload: UserState['token']): void {
    state.token = payload;
  }
};

export const actions = {
  async login(
    { commit, dispatch }: UserActionContext,
    payload: LoginInfo
  ): Promise<boolean> {
    try {
      const response = await fetch(endpoints.signin, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(payload)
      });

      const result = await response.json();

      if (!response.ok) throw new Error(result as string);

      saveAuthTokenInSession(result.token as string);
      commit('setToken', result.token);
      commit('setUserId', result.userId);

      const success = await dispatch('getUser', result.userId as string);
      if (!success) throw new Error('Error');

      return true;
    } catch (error) {
      console.log(error, 'In the error block');
      return false;
    }
  },
  async getUser({ commit }: UserActionContext, payload: string) {
    const requestURL = endpoints.getProfile.replace(':id', payload);
    const token = getAuthTokenInSession();
    try {
      if (!token) throw new Error('No valid token');
      const response = await fetch(requestURL, {
        method: 'GET',
        headers: {
          'Content-Type': 'application/json',
          Authorization: token
        }
      });
      const result = await response.json();
      if (!response.ok) throw new Error(result);

      commit('setUser', result);
      commit('toggleSignIn', true);
      return true;
    } catch (error) {
      console.log(error);
      return false;
    }
  },
  async registerUser(
    { commit }: UserActionContext,
    payload: RegisterInfo
  ): Promise<boolean> {
    try {
      const response = await fetch(endpoints.register, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(payload)
      });

      const result = await response.json();

      if (!response.ok) throw new Error(result as string);

      saveAuthTokenInSession(result.session.token as string);
      commit('setUser', result.register.user);

      return true;
    } catch (error) {
      console.log(error);
      return false;
    }
  },
  getToken({ commit, dispatch }: UserActionContext) {
    const token = getAuthTokenInSession();

    if (token) {
      commit('setToken', token);
      dispatch('authenticated');
    }
  },
  async authenticated({ state, dispatch }: UserActionContext) {
    const token = state.token;

    if (token) {
      try {
        const response = await fetch(endpoints.signin, {
          method: 'POST',
          headers: {
            'Content-Type': 'application/json',
            Authorization: token
          }
        });

        const result = await response.json();

        if (!response.ok) throw new Error(result as string);

        dispatch('getUser', result.id);
      } catch (error) {
        console.log(error);
      }
    }
  }
};

const userModule = {
  namespaced: true,
  state,
  getters,
  mutations,
  actions
};

export default userModule;
