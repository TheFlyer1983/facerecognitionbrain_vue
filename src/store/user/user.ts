import { defineStore } from 'pinia';
import { UpdateInfo, User, UserState } from './userTypes';
import {
  LoginInfo,
  LoginResponse,
  RegisterInfo,
  RegisterResponse
} from '@/types';
import request from '@/functions/request';
import { endpoints } from '@/constants';
import {
  getAuthTokenInSession,
  removeAuthTokenFromSession,
  saveAuthTokenInSession
} from '@/functions/storageFunctions';

export const useUserStore = defineStore('user', {
  state: (): UserState => ({
    isSignedIn: false,
    isProfileOpen: false,
    id: '',
    user: null,
    token: '',
    rank: ''
  }),

  actions: {
    async login(payload: LoginInfo) {
      payload = { ...payload, returnSecureToken: true };

      try {
        const response = await request.post<LoginResponse>(
          endpoints.signin,
          payload,
          {
            params: { key: import.meta.env.VITE_APP_FIREBASE_API_KEY }
          }
        );

        saveAuthTokenInSession(response.data.idToken);
        this.token = response.data.idToken;
        this.id = response.data.localId;

        const success = await this.getUser(response.data.localId);
        if (!success) throw new Error('Error');

        return true;
      } catch (error) {
        console.error(error);
        return false;
      }
    },

    async getUser(userId: string) {
      const requestURL = endpoints.profile.replace(':id', userId);

      try {
        const response = await request.get<User>(requestURL, {
          params: { auth: this.token }
        });

        this.user = response.data;
        this.isSignedIn = true;

        //this.getRank();

        return true;
      } catch (error) {
        console.error(error);
        return false;
      }
    },

    async registerUser(payload: RegisterInfo) {
      const registerPayload = {
        email: payload.email,
        password: payload.password,
        returnSecureToken: true
      };
      try {
        const response = await request.post<RegisterResponse>(
          endpoints.register,
          registerPayload,
          {
            params: { key: import.meta.env.VITE_APP_FIREBASE_API_KEY }
          }
        );

        saveAuthTokenInSession(response.data.idToken);
        this.token = response.data.idToken;
        this.id = response.data.localId;
        await this.createProfile(payload.name);

        return true;
      } catch (error) {
        console.error(error);
        return false;
      }
    },

    async createProfile(name: string) {
      const profilePayload = {
        name,
        entries: 0,
        joined: new Date()
      };

      const requestURL = endpoints.profile.replace(':id', this.id);
      try {
        const response = await request.put(requestURL, profilePayload, {
          params: { auth: this.token }
        });

        this.user = response.data;
      } catch (error) {
        console.log(error);
      }
    },

    getToken() {
      const token = getAuthTokenInSession();

      if (token) {
        this.token = token;

        this.authenticated();
      }
    },

    async authenticated() {
      if (this.token) {
        try {
          const response = await request.post(endpoints.signin);

          this.getUser(response.data.id);
        } catch (error) {
          console.error(error);
        }
      }
    },

    async getRank() {
      const entries = this.user?.entries;

      try {
        const response = await request.post(endpoints.rank, { entries });
        this.rank = response.data;
      } catch (error) {
        console.error(error);
      }
    },

    signout() {
      removeAuthTokenFromSession();
      this.$reset();
    },

    async updateUser(payload: UpdateInfo) {
      const requestURL = endpoints.profile.replace(':id', this.id);
      try {
        await request.put(requestURL, payload);

        this.getUser(this.id as string);
      } catch (error) {
        console.error(error);
      }
    },

    async deleteUser() {
      const requestURL = endpoints.profile.replace(':id', this.id);
      try {
        await request.delete(requestURL, { params: { auth: this.token } });

        await request.post(
          endpoints.delete,
          { idToken: this.id },
          { params: { key: import.meta.env.VITE_APP_FIREBASE_API_KEY } }
        );

        this.signout();
      } catch (error) {
        console.error(error);
      }
    }
  }
});
