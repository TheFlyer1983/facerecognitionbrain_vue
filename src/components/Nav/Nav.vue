<script setup lang="ts">
import { computed } from 'vue';
import { useUserStore } from '@/store/user';
import { useNavigation } from '@/modules/navigation';
import { Routes } from '@/router/routes';
import ProfileIcon from '../Profile/ProfileIcon.vue';

const userStore = useUserStore();
const { navigate } = useNavigation();

const isSignedIn = computed(() => userStore.isSignedIn);

function signOut() {
  userStore.signout();
  navigate({ name: Routes.Login });
}
</script>

<template>
  <nav class="nav">
    <template v-if="isSignedIn">
      <ProfileIcon @signout="signOut" />
    </template>
    <template v-else>
      <p
        class="nav--link"
        @click="navigate({ name: Routes.Login })"
        data-test="signin"
      >
        Sign In
      </p>
      <p
        class="nav--link"
        @click="navigate({ name: Routes.Register })"
        data-test="register"
      >
        Register
      </p>
    </template>
  </nav>
</template>
<style scoped lang="scss">
.nav {
  display: flex;
  justify-content: flex-end;

  &--link {
    font-size: 1.5rem;
    color: black;
    text-decoration: underline;
    padding: 1rem;
    cursor: pointer;
    opacity: 1;
    transition: opacity 0.15s ease-in;
    margin-top: 0;
    margin-bottom: 16px;

    &:hover {
      opacity: 0.5;
      transition: opacity 0.15s ease-in;
    }
  }
}
</style>
