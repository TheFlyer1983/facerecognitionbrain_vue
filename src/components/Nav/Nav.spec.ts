import { reactive } from 'vue';
import { shallowMount, VueWrapper } from '@vue/test-utils';
import { useStore } from 'vuex';
import { useNavigation } from '@/modules/navigation';

import Nav from './Nav.vue';
import { UserState } from '@/store/modules/user/userTypes';
import { Routes } from '@/router/routes';

jest.mock('vuex');
jest.mock('@/modules/navigation');

const dispatchMock = jest.fn();
const navigateMock = jest.fn();

const mockedUseStore = useStore as jest.Mock;
const mockedUseNavigation = useNavigation as jest.Mock;

describe('Given the Nav component', () => {
  const render = () => shallowMount(Nav);
  let wrapper: VueWrapper<InstanceType<typeof Nav>>;

  const mockedGetters = {
    ['user/getIsSignedId']: true as UserState['isSignedIn']
  };

  const storeMock = reactive({
    getters: mockedGetters,
    dispatch: dispatchMock
  });

  beforeAll(() => {
    mockedUseStore.mockReturnValue(storeMock);
    mockedUseNavigation.mockReturnValue({
      navigate: navigateMock
    });
  });

  describe('when it is rendered', () => {
    beforeEach(() => {
      wrapper = render();
    });

    it('should match the snapshot', () => {
      expect(wrapper.element).toMatchSnapshot();
    });

    describe('and when the `signOut` function is called', () => {
      beforeEach(() => {
        // wrapper.findComponent(ProfileIcon).//.vm.$emits('signout');
        wrapper.vm.signOut();
      });

      it('should dispatch the correct action', () => {
        expect(storeMock.dispatch).toHaveBeenCalledWith('user/signOut');
      });

      it('should navigate to the login page', () => {
        expect(navigateMock).toHaveBeenCalledWith({
          name: Routes.Login
        });
      });
    });

    describe('and when the `Sign In` link is clicked', () => {
      beforeEach(() => {
        wrapper.find('[data-test="signin"]').trigger('click');
      });

      it('should navigate to the Login page', () => {
        expect(navigateMock).toHaveBeenCalledWith({
          name: Routes.Login
        });
      });
    });

    describe('and when the `Register` link is clicked', () => {
      beforeEach(() => {
        wrapper.find('[data-test="register"]').trigger('click');
      });

      it('should navigate to the Register page', () => {
        expect(navigateMock).toHaveBeenCalledWith({
          name: Routes.Register
        });
      });
    });
  });
});
