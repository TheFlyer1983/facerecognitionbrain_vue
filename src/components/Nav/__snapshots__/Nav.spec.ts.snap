// Vitest Snapshot v1

exports[`Given the Nav component > when it is rendered > should match the snapshot 1`] = `
<nav
  class="nav"
  data-v-542b3c97=""
>
  
  <p
    class="nav--link"
    data-test="signin"
    data-v-542b3c97=""
  >
     Sign In 
  </p>
  <p
    class="nav--link"
    data-test="register"
    data-v-542b3c97=""
  >
     Register 
  </p>
  
</nav>
`;
